File: smt-pro-catalog/backend/src/services/product.service.js
const prisma = require('../config/prisma');
const { cloudinary } = require('../config/cloudinary');

const getAll = async ({ category, search, page = 1, limit = 20 } = {}) => {
  const where = {};

  if (category) where.category = { equals: category, mode: 'insensitive' };
  if (search) {
    where.OR = [
      { name: { contains: search, mode: 'insensitive' } },
      { description: { contains: search, mode: 'insensitive' } },
    ];
  }

  const skip = (parseInt(page) - 1) * parseInt(limit);
  const take = parseInt(limit);

  const [products, total] = await Promise.all([
    prisma.product.findMany({
      where,
      orderBy: { createdAt: 'desc' },
      skip,
      take,
      select: {
        id: true, name: true, description: true, price: true,
        quantity: true, imageUrl: true, category: true, createdAt: true,
      },
    }),
    prisma.product.count({ where }),
  ]);

  return { products, total, page: parseInt(page), limit: take };
};

const getById = async (id) => {
  const product = await prisma.product.findUnique({
    where: { id: parseInt(id) },
    select: {
      id: true, name: true, description: true, price: true,
      quantity: true, imageUrl: true, category: true, createdAt: true, updatedAt: true,
    },
  });
  if (!product) throw new Error('PRODUCT_NOT_FOUND');
  return product;
};

const create = async ({ name, description, price, quantity, category, imageUrl, imagePublicId }) => {
  return prisma.product.create({
    data: { name, description, price, quantity, category, imageUrl, imagePublicId },
    select: {
      id: true, name: true, description: true, price: true,
      quantity: true, imageUrl: true, category: true, createdAt: true,
    },
  });
};

const update = async (id, updates, newImagePublicId) => {
  const existing = await prisma.product.findUnique({ where: { id: parseInt(id) } });
  if (!existing) throw new Error('PRODUCT_NOT_FOUND');

  if (newImagePublicId && existing.imagePublicId) {
    await cloudinary.uploader.destroy(existing.imagePublicId).catch(() => {});
  }

  return prisma.product.update({
    where: { id: parseInt(id) },
    data: updates,
    select: {
      id: true, name: true, description: true, price: true,
      quantity: true, imageUrl: true, category: true, updatedAt: true,
    },
  });
};

const remove = async (id) => {
  const existing = await prisma.product.findUnique({ where: { id: parseInt(id) } });
  if (!existing) throw new Error('PRODUCT_NOT_FOUND');

  if (existing.imagePublicId) {
    await cloudinary.uploader.destroy(existing.imagePublicId).catch(() => {});
  }

  await prisma.product.delete({ where: { id: parseInt(id) } });
};

module.exports = { getAll, getById, create, update, remove };

File: smt-pro-catalog/flutter_app/lib/features/products/services/product_service.dart
import '../../../core/api/api_client.dart';
import '../../../core/constants/api_endpoints.dart';
import '../models/product_model.dart';

class ProductService {
  final _client = ApiClient();

  Future<List<ProductModel>> getAll({String? search, String? category}) async {
    final res = await _client.get(ApiEndpoints.products, params: {
      if (search   != null && search.isNotEmpty)   'search':   search,
      if (category != null && category.isNotEmpty) 'category': category,
      'limit': 100,
    });
    final data = res.data['data'] as Map<String, dynamic>;
    return (data['products'] as List)
        .map((e) => ProductModel.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<ProductModel> create({
    required String name,
    required double price,
    required int    quantity,
    required String category,
    String?         description,
  }) async {
    final res = await _client.post(ApiEndpoints.products, data: {
      'name':        name,
      'price':       price,
      'quantity':    quantity,
      'category':    category,
      if (description != null && description.isNotEmpty) 'description': description,
    });
    return ProductModel.fromJson(res.data['data'] as Map<String, dynamic>);
  }

  Future<ProductModel> update(int id, Map<String, dynamic> updates) async {
    final res = await _client.put('${ApiEndpoints.products}/$id', data: updates);
    return ProductModel.fromJson(res.data['data'] as Map<String, dynamic>);
  }

  Future<void> delete(int id) async {
    await _client.delete('${ApiEndpoints.products}/$id');
  }
}
